### PR TITLE
add support for cuSolverDN

### PR DESCRIPTION
**Pull request by map9959**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -208,6 +209,47 @@ if(POLYSOLVE_WITH_SPECTRA)
     target_compile_definitions(polysolve PUBLIC -DPOLYSOLVE_WITH_SPECTRA)
 endif()
 
+# cuSolver solvers
+if(POLYSOLVE_WITH_CUSOLVER)
```
Comment (teseoch) [inline CMakeLists.txt:213]: this thing should go in a recpipe

File: cmake/recipes/spdlog.cmake [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#
```
Comment (teseoch) [inline cmake/recipes/spdlog.cmake:1]: we don't want spdlog

File: src/polysolve/LinearSolver.hpp [hunk 3]
```diff
@@ -85,6 +85,14 @@ namespace polysolve
         // Factorize system matrix
         virtual void factorize(const StiffnessMatrix &A) {}
 
+        #ifdef POLYSOLVE_WITH_CUSOLVER
```
Comment (teseoch) [inline src/polysolve/LinearSolver.hpp:88]: no ifdef, we can just have overload for the dense solvers

File: src/polysolve/LinearSolverCuSolverDN.cu [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+////////////////////////////////////////////////////////////////////////////////
+#include <polysolve/LinearSolverCuSolverDN.cuh>
+#include <fstream>
+#include <string>
+#include <iostream>
+
+#include "spdlog/spdlog.h"
+
+#define gpuErrchk(ans){ gpuAssert((ans), __FILE__, __LINE__); }
```
Comment (teseoch) [inline src/polysolve/LinearSolverCuSolverDN.cu:11]: this is bad, do you need it?

Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cu:11]: I thought it was common practice to check GPU code for errors by putting each CUDA code line in a function like this

File: src/polysolve/LinearSolverCuSolverDN.cu [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    void LinearSolverCuSolverDN::analyzePattern(const Eigen::MatrixXd &A, const int precond_num)
+    {
+
+    }
+
+    void LinearSolverCuSolverDN::factorize(const StiffnessMatrix &A)
+    {
+        numrows = (int)A.rows();
+
+        Adense = Eigen::MatrixXd(A);
```
[the remark this answers is not in the document]
Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cu:65]: there are two factorize methods, this is for if a sparse matrix is inputted (it's converted to a dense matrix), the dense factorize method is the same without the conversation part

File: CMakeLists.txt (hunk 1 above)
Comment (map9959) [reply CMakeLists.txt:213]: created a recipe

File: cmake/recipes/spdlog.cmake (hunk 2 above)
Comment (map9959) [reply cmake/recipes/spdlog.cmake]: removed spdlog

File: src/polysolve/LinearSolver.hpp (hunk 3 above)
Comment (map9959) [reply src/polysolve/LinearSolver.hpp]: removed ifdef section

File: src/polysolve/LinearSolverCuSolverDN.cu [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+#include <iostream>
+
+#include "spdlog/spdlog.h"
+
+#define gpuErrchk(ans){ gpuAssert((ans), __FILE__, __LINE__); }
+
+inline void gpuAssert(cudaError_t code, const char *file, int line,
+                      bool abort = true) {
+  if (code != cudaSuccess) {
+    spdlog::error("GPUassert: {} {} {:d}\n", cudaGetErrorString(code), file,
```
[the remark this answers is not in the document]
Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cu]: It now throws an exception, I don't know if there's a more "standard" format for one

File: src/polysolve/LinearSolverCuSolverDN.cu [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        gpuErrchk(cudaMemcpy(d_b, (const void *)b.data(), sizeof(double) * b.size(), cudaMemcpyHostToDevice));
+
+        //solve
+        cusolverStatus_t solvererr = cusolverDnXgetrs(cuHandle, cuParams, CUBLAS_OP_N, numrows, 1,
+            CUDA_R_64F, d_A, numrows, d_Ipiv,
+            CUDA_R_64F, d_b, numrows, d_info);
+        if(solvererr == CUSOLVER_STATUS_SUCCESS){
+            //std::cout << "success" << std::endl;
+        }else if(solvererr == CUSOLVER_STATUS_INVALID_VALUE){
+            std::cout << "invalid value" << std::endl;
```
[the remark this answers is not in the document]
Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cu]: made into exceptions

File: src/polysolve/LinearSolverCuSolverDN.cu [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+#include "spdlog/spdlog.h"
+
+#define gpuErrchk(ans){ gpuAssert((ans), __FILE__, __LINE__); }
+
+inline void gpuAssert(cudaError_t code, const char *file, int line,
+                      bool abort = true) {
+  if (code != cudaSuccess) {
+    spdlog::error("GPUassert: {} {} {:d}\n", cudaGetErrorString(code), file,
+                  line);
+    if (abort)
```
[the remark this answers is not in the document]
Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cu]: made into an exception

File: src/polysolve/LinearSolverCuSolverDN.cu (hunk 4 above)
Comment (teseoch) [reply src/polysolve/LinearSolverCuSolverDN.cu:9]: macros are bad, we can have a function

File: src/polysolve/LinearSolverCuSolverDN.cuh [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        size_t h_lwork = 0;     /* size of workspace */
+        void *h_work = nullptr; /* host workspace for getrf */
+        int *d_info = nullptr; //error info
+
+        int numrows;
+
+        Eigen::MatrixXd Adense;
+    };
+
+//TODO: complete this -MP
```
Comment (teseoch) [inline src/polysolve/LinearSolverCuSolverDN.cuh:88]: is this a real todo?

File: tests/test_solver.cpp [hunk 10]
```diff
@@ -571,3 +571,102 @@ TEST_CASE("amgcl_blocksolver_crystm03_Bicgstab", "[solver]")
...
+    solver->analyzePattern(A, A.rows());
+    solver->factorize(A);
+    solver->solve(b, x);
+
+    // std::cout<<"Solver error: "<<x<<std::endl;
+    const double err = (A * x - b).norm();
+    REQUIRE(err < 1e-8);
+}
+/*
+TEST_CASE("cusolverdn_5cubes", "[solver]")
```
Comment (teseoch) [inline tests/test_solver.cpp:625]: why is this off?

File: src/polysolve/LinearSolverCuSolverDN.cu [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    {
+        cusolverDnCreate(&cuHandle);
+        cusolverDnCreateParams(&cuParams);
+        cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking);
+        cusolverDnSetStream(cuHandle, stream);
+    }
+
+    void LinearSolverCuSolverDN::setParameters(const json &params)
+    {
+        //TODO: make a json option for legacy solver vs new solver
```
Comment (teseoch) [inline src/polysolve/LinearSolverCuSolverDN.cu:36]: if it dosent have options, no need of this function

Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cu:36]: i was going to include a choice for legacy vs. new solver but i can't see a use case for the legacy one

File: tests/test_solver.cpp (hunk 10 above)
Comment (map9959) [reply tests/test_solver.cpp:625]: it requires additional data that isn't included in the polysolve data folder (the gradients and hessians from the 5cubes simulation on polyfem), so i commented it out

File: src/polysolve/LinearSolverCuSolverDN.cu (hunk 4 above)
Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cu:9]: made it a function

File: src/polysolve/LinearSolverCuSolverDN.cu (hunk 5 above)
[the remark this answers is not in the document]
Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cu:60]: oops! did that

File: CMakeLists.txt [hunk 12]
```diff
@@ -208,6 +209,13 @@ if(POLYSOLVE_WITH_SPECTRA)
...
+    include(cusolverdn)
+    target_link_libraries(polysolve PUBLIC CUDA::cusolver)
```
[the remark this answers is not in the document]
Comment (map9959) [reply CMakeLists.txt]: fixed

File: src/polysolve/LinearSolverCuSolverDN.cu [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+////////////////////////////////////////////////////////////////////////////////
+#include <polysolve/LinearSolverCuSolverDN.cuh>
+#include <fstream>
+#include <string>
+#include <iostream>
+
+#define gpuErrchk(ans){ gpuAssert((ans), __FILE__, __LINE__); }
+
+inline void gpuAssert(cudaError_t code, const char *file, int line,
+                      bool abort = true) {
```
[the remark this answers is not in the document]
Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cu]: no, fixed it

File: src/polysolve/LinearSolverCuSolverDN.cuh (hunk 9 above)
Comment (map9959) [reply src/polysolve/LinearSolverCuSolverDN.cuh]: no, removed it. i'll be more clear in the future (for the benefit of myself a couple weeks down the line)

File: tests/test_solver.cpp (hunk 10 above)
Comment (teseoch) [reply tests/test_solver.cpp:625]: i can push the data to the repo, so we can have it

**Review by teseoch — APPROVED** [review]

**Pull request merged**